cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java

```diff
@@ -110,6 +114,7 @@
   private final String hostname;
   private Callable<ListenableFuture<RunId>> submitSpark;
   private Runnable cleanupTask;
+  private volatile boolean stopRequested;
```

Comment:
No need to add this variable. In the `run` method, if the `jobCompletion` is cancelled, then stop is requested. Using an extra variable actually introducing a race. You can just set the program state in the `run()` method based on the state in the `jobCompletion` future.